ports/esp32s2/common-hal/analogio/AnalogIn.c

```diff
@@ -0,0 +1,98 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "common-hal/analogio/AnalogIn.h"
+#include "py/mperrno.h"
+#include "py/runtime.h"
+#include "supervisor/shared/translate.h"
+
+#include "components/driver/include/driver/adc_common.h"
+#include "components/esp_adc_cal/include/esp_adc_cal.h"
+
+#include "shared-bindings/microcontroller/Pin.h"
+
+#define DEFAULT_VREF        1100
+#define NO_OF_SAMPLES       64
+#define ATTENUATION         ADC_ATTEN_DB_11
+#define DATA_WIDTH          ADC_WIDTH_BIT_13
+
+void common_hal_analogio_analogin_construct(analogio_analogin_obj_t* self,
+        const mcu_pin_obj_t *pin) {
+    if (pin->adc_index == 0 || pin->adc_channel == ADC_CHANNEL_MAX) {
+        mp_raise_ValueError(translate("Pin does not have ADC capabilities"));
+    }
+    common_hal_mcu_pin_claim(pin);
+    self->pin = pin;
+}
+
+bool common_hal_analogio_analogin_deinited(analogio_analogin_obj_t *self) {
+    return self->pin == NULL;
+}
+
+void common_hal_analogio_analogin_deinit(analogio_analogin_obj_t *self) {
+    if (common_hal_analogio_analogin_deinited(self)) {
+        return;
+    }
+    reset_pin_number(self->pin->number);
+    self->pin = NULL;
+}
+
+uint16_t common_hal_analogio_analogin_get_value(analogio_analogin_obj_t *self) {
+    if (self->pin->adc_index == ADC_UNIT_1) {
+        adc1_config_width(DATA_WIDTH);
+        adc1_config_channel_atten((adc1_channel_t)self->pin->adc_channel, ATTENUATION);
+    } else if (self->pin->adc_index == ADC_UNIT_2) {
+        adc2_config_channel_atten((adc2_channel_t)self->pin->adc_channel, ATTENUATION);
+    }
+
+    // Automatically select calibration process depending on status of efuse
+    esp_adc_cal_characteristics_t *adc_chars = calloc(1, sizeof(esp_adc_cal_characteristics_t));
+    esp_adc_cal_characterize(self->pin->adc_index, ATTENUATION, DATA_WIDTH, DEFAULT_VREF, adc_chars);
+
+    uint32_t adc_reading = 0;
+    //Multisampling
+    for (int i = 0; i < NO_OF_SAMPLES; i++) {
+        if (self->pin->adc_index == ADC_UNIT_1) {
+            adc_reading += adc1_get_raw((adc1_channel_t)self->pin->adc_channel);
+        } else {
+            int raw;
+            esp_err_t r = adc2_get_raw((adc2_channel_t)self->pin->adc_channel, DATA_WIDTH, &raw);
+            if ( r != ESP_OK ) {
+                mp_raise_ValueError(translate("ADC2 is being used by WiFi"));
```

Comment:
Is there a specific error code for this? I thought the S2 would automatically share the ADC with WiFi.

Reply:
It does share, to the extent that it has a hardware arbitrator that lets the user access the ADC2 as long as Wifi isn't using it, as compared to the ESP32 which I think just locks down all of ADC2 completely from the moment you start WiFi. But there will still be moments where you can't access it, users will have to use a Try/Except to get around it.